deepchecks/vision/checks/performance/performance_report.py

```diff
@@ -45,44 +45,56 @@ def __init__(self,
         super().__init__()
         self.alternative_metrics = alternative_metrics
         self.prediction_formatter = prediction_formatter
+        self._state = {}
 
-    def run_logic(self, context: Context) -> CheckResult:
-        """Run check.
-
-        Returns
-        -------
-        CheckResult
-            value is dictionary in format 'score-name': score-value
-        """
-        train_dataset = context.train
-        test_dataset = context.test
-        model = context.model
+    def initialize_run(self, context: Context):
+        """Optional method to initialize run before starting updating on batches."""
+        self._state = {}
         context.assert_task_type(TaskType.CLASSIFICATION, TaskType.OBJECT_DETECTION)
-
         # Get default scorers if no alternative, or validate alternatives
-        scorers = get_scorers_list(test_dataset, self.alternative_metrics)
-        datasets = {'Train': train_dataset, 'Test': test_dataset}
-
-        classes = train_dataset.get_samples_per_class().keys()
-        plot_x_axis = 'Class'
+        self._state['train'] = {}
+        self._state['test'] = {}
+        self._state['train']['scorers'] = get_scorers_list(context.train, self.alternative_metrics)
+        self._state['test']['scorers'] = get_scorers_list(context.train, self.alternative_metrics)
+        for dataset_name in ['train', 'test']:
+            for _, metric in self._state[dataset_name]['scorers'].items():
+                metric.reset()
+        self._state['train']['n_samples'] = context.train.get_samples_per_class()
+        self._state['test']['n_samples'] = context.test.get_samples_per_class()
+
+        self._state['classes'] = context.train.get_samples_per_class().keys()
+
+    def update(self, context: Context, batch: Any, dataset_name: str = 'train'):
+        """Update internal check state with given batch for either train or test."""
+        if dataset_name == 'train':
+            dataset = context.train
+        else:
+            dataset = context.test
+        images = batch[0]
+        label = dataset.label_transformer(batch[1])
+        prediction = self.prediction_formatter(context.infer(images))
+        for metric_name, metric in self._state[dataset_name]['scorers'].items():
+            metric.update((prediction, label))
+
+    def compute(self, context: Context) -> CheckResult:
+        """Compute final check result based on accumulated internal state."""
         results = []
-
-        for dataset_name, dataset in datasets.items():
-            n_samples = dataset.get_samples_per_class()
+        for dataset_name in ['train', 'test']:
+            n_samples = self._state[dataset_name]['n_samples']
             results.extend(
                 [dataset_name, class_name, name, class_score, n_samples[class_name]]
-                for name, score in calculate_metrics(list(scorers.values()), dataset, model,
```

Comment:
do we need the calculate_metrics then? maybe we should delete it?

Reply:
Done